protocol/payment.go

```diff
@@ -0,0 +1,510 @@
+	case Iden3PaymentRailsERC20RequestV1Type:
+		var o Iden3PaymentRailsERC20RequestV1
+		if err := json.Unmarshal(data, &o); err != nil {
+			return fmt.Errorf("unmarshalling PaymentRequestInfoData: %w", err)
+		}
+		p.railsERC = []Iden3PaymentRailsERC20RequestV1{o}
+	default:
+		return errors.Errorf("unmarshalling PaymentRequestInfoData. unknown type: %s", typ)
+	}
+	return nil
+}
+
+func (p *PaymentRequestInfoData) unmarshalFromCollection(typ string, data []byte) error {
+	switch typ {
+	case Iden3PaymentRequestCryptoV1Type:
+		if err := json.Unmarshal(data, &p.crypto); err != nil {
+			return fmt.Errorf("unmarshalling PaymentRequestInfoData: %w", err)
+		}
+	case Iden3PaymentRailsRequestV1Type:
+		if err := json.Unmarshal(data, &p.rails); err != nil {
+			return fmt.Errorf("unmarshalling PaymentRequestInfoData: %w", err)
+		}
+	case Iden3PaymentRailsERC20RequestV1Type:
+		if err := json.Unmarshal(data, &p.railsERC); err != nil {
+			return fmt.Errorf("unmarshalling PaymentRequestInfoData: %w", err)
+		}
+	default:
+		return errors.Errorf("unmarshalling PaymentRequestInfoData. unknown type: %s", typ)
+	}
+	return nil
+}
+
+// Iden3PaymentRequestCryptoV1 represents the Iden3PaymentRequestCryptoV1 payment request data.
+type Iden3PaymentRequestCryptoV1 struct {
+	Type       string `json:"type"`
+	ID         string `json:"id"`
+	Context    string `json:"@context,omitempty"`
+	ChainID    string `json:"chainId"`
+	Address    string `json:"address"`
+	Amount     string `json:"amount"`
+	Currency   string `json:"currency"`
+	Expiration string `json:"expiration,omitempty"`
+}
+
+// Iden3PaymentRailsRequestV1 represents the Iden3PaymentRailsRequestV1 payment request data.
+type Iden3PaymentRailsRequestV1 struct {
+	Nonce          string                         `json:"nonce"`
+	Type           string                         `json:"type"`
+	Context        PaymentContext                 `json:"@context"`
+	Recipient      string                         `json:"recipient"`
+	Amount         string                         `json:"amount"` // Not negative number
+	ExpirationDate string                         `json:"expirationDate"`
+	Proof          EthereumEip712Signature2021Col `json:"proof"`
+	Metadata       string                         `json:"metadata"`
+	Currency       string                         `json:"currency"`
+}
+
+// Iden3PaymentRailsERC20RequestV1 represents the Iden3PaymentRailsERC20RequestV1 payment request data.
+type Iden3PaymentRailsERC20RequestV1 struct {
+	Nonce          string                         `json:"nonce"`
+	Type           string                         `json:"type"`
+	Context        PaymentContext                 `json:"@context"`
+	Recipient      string                         `json:"recipient"`
+	Amount         string                         `json:"amount"` // Not negative number
+	ExpirationDate string                         `json:"expirationDate"`
+	Proof          EthereumEip712Signature2021Col `json:"proof"`
```

Comment:
I have one doubt regarding that. From the protocol persperctive we can have different proof types here. So we can design structure not to break interfaces later.

What do you think @x1m3 @volodymyr-basiuk ?

Reply:
also it can be an object, not always an array

Reply:
Yes. I agree.  We need to handle objects and arrays and change Proof structure in a way that can be extended in the future.

Reply:
Fixed. I changed it to class that can be extended in the future with other proof types.